src/EventStore.Core/DataStructures/ProbabilisticFilter/FileStreamPersistence.cs

```diff
@@ -0,0 +1,275 @@
+				_path,
+				FileMode.Open,
+				FileAccess.Read,
+				FileShare.ReadWrite,
+				bufferSize: 65_536,
+				options: FileOptions.SequentialScan);
+
+			//qq reliable check?
+			if (bulkFileStream.Length != DataAccessor.FileSize)
+				throw new SizeMismatchException(
+					$"The expected file size ({DataAccessor.FileSize:N0}) does not match " +
+					$"the actual file size ({bulkFileStream.Length:N0}) of file {_path}");
+
+			var bytesToRead = DataAccessor.FileSize;
+			var bytesRead = 0L;
+			while (bytesToRead > 0) {
+				var bytesToReadInBlock = bytesToRead > int.MaxValue
+					? int.MaxValue
+					: (int)bytesToRead;
+
+				bulkFileStream.Read(new Span<byte>(DataAccessor.Pointer + bytesRead, bytesToReadInBlock));
+
+				bytesRead += bytesToReadInBlock;
+				bytesToRead -= bytesToReadInBlock;
+			}
+
+			var elapsed = sw.Elapsed;
+			var fileSizeMb = DataAccessor.FileSize / 1000 / 1000;
+			var megaBytesPerSecond = fileSizeMb / elapsed.TotalSeconds;
+			Log.Information(
+				"Read persisted bloom filter {path} into memory. Took {elapsed}. {megaBytesPerSecond} MB/s",
+				_path,
+				elapsed,
+				megaBytesPerSecond);
+		}
+
+		private void OnPageDirty(long pageNumber) {
+			lock (_bitmapLock) {
+				ThrowIfDisposed();
+				var byteIndex = (int)(pageNumber / 8);
+				var bitIndex = pageNumber % 8;
+				ref var byteValue = ref _dirtyPageBitmap.AsSpan()[byteIndex];
+				byteValue = byteValue.SetBit(bitIndex);
+			}
+		}
+
+		public void Flush() {
+			using var fileStream = new FileStream(
+				_path,
+				FileMode.OpenOrCreate,
+				FileAccess.ReadWrite,
+				FileShare.ReadWrite,
+				bufferSize: DataAccessor.PageSize);
+
+			fileStream.SetLength(DataAccessor.FileSize);
+
+			Span<byte> localCacheLine = stackalloc byte[BloomFilterIntegrity.CacheLineSize];
+			localCacheLine.Clear();
+
+			var pageNumber = 0;
+			var flushedPages = 0;
+			var pauses = 0;
+
+			//qq configurable?
+			var flushBatchSize = 512;
+			var flushBatchDelay = 100; // enough to get out of the way of any other writes
+			var activelyFlushing = Stopwatch.StartNew();
+
+			for (var remaining = _dirtyPageBitmap.AsSpan();
+				remaining.Length > 0;
+				remaining = remaining[BloomFilterIntegrity.CacheLineSize..]) {
+
+				lock (_bitmapLock) {
+					ThrowIfDisposed();
+					var cacheLine = remaining[..BloomFilterIntegrity.CacheLineSize];
+					cacheLine.CopyTo(localCacheLine);
+					cacheLine.Clear();
+				}
+
+				foreach (var @byte in localCacheLine) {
+					for (var bitOffset = 0; bitOffset < 8; bitOffset++) {
+						if (@byte.IsBitSet(bitOffset)) {
+							WritePage(pageNumber, fileStream);
+							flushedPages++;
+
+							if (flushedPages % flushBatchSize == 0) {
+								activelyFlushing.Stop();
+								pauses++;
+								Thread.Sleep(flushBatchDelay);
+								activelyFlushing.Start();
+							}
+						}
+
+						pageNumber++;
+						if (pageNumber == DataAccessor.NumPages)
+							goto Done;
+					}
+				}
+			}
+
+
+			Done:
+			fileStream.FlushToDisk();
```

Comment:
may be worth considering flushing in small batches instead of doing one big flush: see https://github.com/EventStore/EventStore/pull/1918

Reply:
woops yes i meant to add a call to flushtodisk in the 'pause'

Reply:
@shaan1337 do you mean to flush(false) multiple times then flush(true) at the end?